File: script/board_at_ply.rb
# Usage:
#   bin/rails runner script/board_at_ply.rb MATCH_ID PLY [--perspective=white|black]

require 'json'
require 'open3'
require 'tempfile'

match_id = ARGV.shift
ply = ARGV.shift&.to_i
perspective = 'white'

abort 'Usage: bin/rails runner script/board_at_ply.rb MATCH_ID PLY [--perspective=white|black]' if match_id.blank? || ply.nil? || ply <= 0

ARGV.each do |arg|
  case arg
  when /\A--perspective=(white|black)\z/
    perspective = Regexp.last_match(1)
  else
    abort "Unknown option: #{arg}"
  end
end

match = Match.find(match_id)
notation = Array(match.movement_notation)

abort "Ply #{ply} is out of range for match #{match.id} (#{notation.length} plies)" if ply > notation.length + 1

payload = {
  id: match.id,
  white: match.white_player.name,
  black: match.black_player.name,
  result: match.result,
  movement_notation: notation,
  ply: ply,
  perspective: perspective
}

Tempfile.create(['board-at-ply', '.json']) do |tempfile|
  tempfile.write(JSON.pretty_generate(payload))
  tempfile.flush

  command = [
    Rails.root.join('node_modules/.bin/vite-node').to_s,
    '--config',
    Rails.root.join('vitest.config.js').to_s,
    Rails.root.join('script/board_at_ply.mjs').to_s,
    tempfile.path
  ]

  stdout, stderr, status = Open3.capture3(*command)
  print stdout
  warn stderr if stderr.present?
  exit(status.exitstatus || 1) unless status.success?
end

File: script/board_at_ply.mjs
import fs from 'fs'

import { boardAscii, buildBoardFromNotationPrefix } from './match_analysis_common.mjs'

const payloadPath = process.argv[2]

if (!payloadPath) {
  console.error('Usage: node board_at_ply.mjs PAYLOAD_PATH')
  process.exit(1)
}

const payload = JSON.parse(fs.readFileSync(payloadPath, 'utf8'))
const prefix = payload.movement_notation.slice(0, payload.ply - 1)
const board = buildBoardFromNotationPrefix(prefix)
const previousMove = payload.ply > 1 ? payload.movement_notation[payload.ply - 2] : null
const nextMove = payload.ply <= payload.movement_notation.length ? payload.movement_notation[payload.ply - 1] : null

console.log(`MATCH ${payload.id} ${payload.white} vs ${payload.black} ${payload.result}`)
console.log(`Board before ply ${payload.ply}`)
console.log(`Allowed to move: ${board.allowedToMove}`)
console.log(`Previous move: ${previousMove || 'none'}`)
console.log(`Next move: ${nextMove || 'none (end of game)'}`)
console.log('')
console.log(boardAscii({ board, perspective: payload.perspective }))

